master/internal/api/handler.go

```diff
@@ -3,13 +3,54 @@ package api
 import (
 	"fmt"
 	"net/http"
+	"net/url"
+	"unicode/utf8"
 
 	"github.com/gorilla/websocket"
 	"github.com/labstack/echo/v4"
 	"github.com/pkg/errors"
 )
 
-var upgrader = websocket.Upgrader{}
+// equalASCIIFold returns true if s is equal to t with ASCII case folding as
+// defined in RFC 4790.
+func equalASCIIFold(s, t string) bool {
+	for s != "" && t != "" {
+		sr, size := utf8.DecodeRuneInString(s)
+		s = s[size:]
+		tr, size := utf8.DecodeRuneInString(t)
+		t = t[size:]
+		if sr == tr {
+			continue
+		}
+		if 'A' <= sr && sr <= 'Z' {
+			sr = sr + 'a' - 'A'
+		}
+		if 'A' <= tr && tr <= 'Z' {
+			tr = tr + 'a' - 'A'
+		}
+		if sr != tr {
+			return false
+		}
+	}
+	return s == t
+}
+
+// Reference: https://github.com/gorilla/websocket/blob/main/server.go#L152
```

Comment:
sorry, but I don't understand how the line here is related, can you share more detail?

Also, you should use a permalink rather than a line number on the main branch, because this link will rot.

Reply:
Of course, so when we don't pass a `checkOrigin` function, it defaults to use `checkSameOrigin` [function](https://github.com/gorilla/websocket/blob/695e9095ce8736ac99c83939ca6b0fe93768f680/server.go#L91) which I copied with a slight modification to allow `localhost`

Reply:
Ah.

Please be much more explicit that you copy/pasted the code in question.  "Reference" could mean lots of things; I didn't realize that this is what it meant.